Add latest releases and a script to get the commit hashes.

daxgames opened the pull request:
I have tested the added hashes and they result in a working pacman install with 2.52.0.1

I also have a `install_pacman.sh` script that works but it does not make sense to add it to this repo or if it does to you I can add a `mcgitty` version of it here also.  We are considering adding pacman to Cmder as a default component. 

See: https://github.com/cmderdev/cmder/pull/3060 

See usage in edited `README.md`

daxgames committed:
Add update.ps1

daxgames committed:
Add update.ps1 generated commit hashes

daxgames committed:
filter results based on -version arg

daxgames committed:
Powershell 5 fixes and cmd.exe shim

daxgames committed:
add 2.47.0.1 to versions

mcgitty commented:
Sorry for the delay. Let me have a look in the coming days. However, since this is for git-bash, it makes more sense to have an automated script in bash instead of Windows CMD.

daxgames commented:
The cmd script is just a shim for a powershell script that does all the work.

Honestly you could do the same and write a `update.sh` that is a shim for the powershell much quicker than rewriting the whole thing in a shell script.

daxgames committed:
add shims

daxgames committed:
add shims, and dos2unix

daxgames committed:
dos2unix

daxgames commented:
Added an `update.sh` shim so the `update.ps1` can be run from `cmd.exe`, `powershell.exe`, `pwsh.exe` or `git bash`.

I chose to write this in `Powershell` because doing it in bash required several external dependencies.

daxgames committed:
x

daxgames committed:
Remove shim files.

mcgitty commented:
I suggest testing your code more. In your `update-script` branch, this failed:
`update.cmd -Version 2.51.2`

I happened to test the following command in cygwin, and the prompt "Enter index of release to use (empty to abort):" was not displayed:
`. update.sh -Version 2.51.0`

daxgames commented:
I do not use Cygwin so I do not have it installed.

I would not expect:

```
. update.sh -Version 2.51.0
```

To display anything.  Not sure I would ever have thought of testing it that way.

But

```
./update.sh -Version 2.51.0

```

Has been tested from git bash and works as expected.

daxgames commented:
I do not use Cygwin so I do not have it installed.

I would not expect:

```
. update.sh -Version 2.51.0
```

To display anything.  Not sure I would ever have thought of testing it that way.

But

```
./update.sh -Version 2.51.0

```

Has been tested from git bash and works as expected.

daxgames commented:
I will look at the other.

daxgames commented:
The problem is with the shim files and not the ps1 script, so I removed the shims.

mcgitty commented:
Do the following from command window cmd.exe:
`powershell -ExecutionPolicy Bypass`
`.\update.ps1 -Version 2.51.2`
The property 'Count' cannot be found on this object. Verify that the property exists.
At C:\Users\michael\git\daxgames-cad0a21\update.ps1:66 char:9
`+`     if ($candidates.Count -gt 1) {
`+`         ~~~~~~~~~~~~~~~~~~~~~~~
`    +` CategoryInfo          : NotSpecified: (:) [], PropertyNotFoundException
`    +` FullyQualifiedErrorId : PropertyNotFoundStrict

daxgames committed:
fix multi vs. single results bug, menu starts at 1, add back shims

daxgames commented:
Thanks for pushing for a fix.  I was busy on other things last time I looked at it and there indeed was a bug when only one candidate was returned.

I also:

1. Added back the shims and tested them.
2. Started menu numbering at `1` when multiple candidates are returned.

In my testing it works as expected now.

daxgames commented:
Never mind I think I see one more issue.  Will let you know when it is right.

daxgames committed:
fix version long when multiple candidates are presented in a manue

daxgames committed:
Exclude '-rc', and work with single/multiple matched releases

daxgames commented:
Should be fixed now.  Please let me know.

daxgames committed:
Update the README.md

daxgames committed:
Update the README.md

mcgitty commented:
Thanks for the update. I have my minimal but concise style of wording, so I will likely change text when I do merge. However, my tests focus on: 1) select the commit ID the same way as the manual-way; 2) the script doesn't crash if matching versions do exist.

For focus 1), the easy way is to find an existing release in the version file and see if the script outputs the same commit ID. So far it is working for a few old versions I tried, but I will test more, including those not in the version file.

For focus 2), I still encounter crashes when it shouldn't. Here are some examples (always select 1):
`update.cmd -Version 2.39`
`update.cmd -Version 2.45.2`

Seeing errors from above cases, I want to suggest a change. Your script seems to print the most recent version first for multiple match. Please make the default selection 1. My intend for the version file is not to capture all possible releases, but only the highest revision of "n.mm.x.y" at the time. Also, the script is non-destructive, so idea of "Abort" is not necessary.

daxgames commented:
@mcgitty We have conflicting goals but I can accomadate and make it work both ways.  I can add  `-latest` switch that will alway output the hashes for what would be  selection #1.  I can make the menu default selection be #1.

My testing always uses a three part semantic version and I was using `2.39.0` where you were using `2.39`, I can even make that work.  

I did get different hashes than in the file for some of the `2.39.0` versions but I attributed it to the manual way being, manual and more prone to human error since every other case I tested returned the same result as in the file.

I never did anything with `2.45.2` but I will look.

daxgames committed:
'-version n.[n].[n]' and '-latest false' are now optional, no args returns latest

daxgames committed:
renamed update.* to get-versiontags.*, add command line help and '-help' arg.

daxgames commented:
@mcgitty I think this is as close to what we both are looking for.  I renamed `updates.*` to `get-versiontags.*` to both match powershell naming standards and `version-tags.txt` name.  All shims also renamed, refactored and tested.

1. Default behavior is to always show latest and not show a menu unless `-latest:$false` is specified.
2. `-version [v]n.[n].[n]` is now optional.
    - If omitted and `-latest:$false` is specified then a menu of ALL releases with valid `package-version-*.txt` files in `build-extra` is displayed.
3. I also added `-help` arg and command line help.

Please lt me know if you see anything else.

daxgames committed:
x

daxgames committed:
Add cmdletbinding and -all

daxgames committed:
README.md updates

daxgames commented:
There is now an `-all` command line option.  See below:

```
PS C:\Users\vagrant\src\github\mcgitty\pacman-for-git> .\get-versiontags -all >x.txt   
Please wait, gathering data for all releases...
PS C:\Users\vagrant\src\github\mcgitty\pacman-for-git> bcomp .\x.txt .\version-tags.txt
```

x64:

<img width="1577" height="729" alt="{127CC6E9-1B08-4DAB-ABEA-4845427E93EA}" src="https://github.com/user-attachments/assets/f1ac50a9-16da-4218-ac34-85df6809a0f9" />

x86:

<img width="1440" height="704" alt="{F9F3BD59-BAFD-4B0E-B0EF-61ADE75729E8}" src="https://github.com/user-attachments/assets/024d889a-98bc-44f5-ad00-7df00fc95bb7" />

daxgames commented:
I think I am done.

daxgames committed:
cleanup

daxgames commented:
Still not happy cleaning it up.

daxgames committed:
Use the same logic for one and all

daxgames commented:
Now I'm done

daxgames committed:
Latest to String

daxgames committed:
README.md changes

daxgames commented:
The git team messed up the latest release by listing the wrong version in the `package-versions.txt`:

<img width="833" height="178" alt="image" src="https://github.com/user-attachments/assets/7b7750e1-df12-4d3f-a206-fb5a22158aaa" />

daxgames committed:
add verbose-outpuut

daxgames commented:
The latest version:

Get a single specific version:
<img width="956" height="274" alt="image" src="https://github.com/user-attachments/assets/5127189d-16b2-4105-ba24-6fa60a99f8a7" />

Get the latest version:

<img width="1155" height="291" alt="image" src="https://github.com/user-attachments/assets/62304ecd-d0eb-4e23-b0d4-b2474ccd9ffe" />

Get all versions:

<img width="1161" height="500" alt="image" src="https://github.com/user-attachments/assets/72e5ecbf-baef-4d84-8e90-f1e8d5f06d2e" />

If a `gh` not avail or `gh auth token` does not return a token:

<img width="1162" height="138" alt="image" src="https://github.com/user-attachments/assets/8d50a2f6-2b00-445b-86ad-a9aa15560b49" />

daxgames committed:
validate package-versions contains the right version and add verbose out for debugging

daxgames committed:
-add outfile flag

daxgames commented:
Added `-y` for auto approve using `gh` supplied token and  `-outfile` switch:

<img width="610" height="442" alt="image" src="https://github.com/user-attachments/assets/3259da43-99ae-4282-a607-f8805edcc7c7" />
<img width="600" height="192" alt="image" src="https://github.com/user-attachments/assets/542b22bf-7816-4adb-aca4-7144e0828ef4" />

daxgames commented:
Now I am done.

daxgames commented:
Have you noticed the latest releases do not seem to have the same versioning in the `package-versions-n.n.n.txt`

<img width="414" height="250" alt="{B0D946CC-5C44-4B42-8A2D-C3CA2B6D67E8}" src="https://github.com/user-attachments/assets/35a14e20-b43b-4ab2-b80b-31743a1167d2" />

Enther the version numbers do not match or missing the commit hash/buildinfo bits.

daxgames commented:
The script still works even though they seem to have changed the `package-versions-n.n.n.txt` versioning.

See: https://github.com/daxgames/pacman-for-git/blob/main/version-tags.txt

<img width="721" height="236" alt="{D1E09CE6-950E-4B9F-82ED-938BD7562AFB}" src="https://github.com/user-attachments/assets/a82c2301-175f-4975-be85-c669aa5400fe" />

daxgames committed:
add output message
